compiler/src/iree/compiler/Codegen/Common/GPU/GPUCombineValueBarriers.cpp

```diff
@@ -0,0 +1,268 @@
+  }
+  auto combinedBarrierOp =
+      rewriter.create<IREE::GPU::ValueBarrierOp>(loc, barrierOperands);
+
+  // Replace all uses of the previous barrier with new barrier.
+  int resultNumber = 0;
+  for (auto barrierOp : valueBarriers) {
+    int numResults = barrierOp.getNumResults();
+    rewriter.replaceOp(barrierOp, combinedBarrierOp->getResults().slice(
+                                      resultNumber, numResults));
+    resultNumber += numResults;
+  }
+  return success();
+}
+
+/// Given two barriers, barrierA and barrierB, combine them into a single
+/// barrier.
+static FailureOr<IREE::GPU::ValueBarrierOp>
+combineValueBarrierPair(RewriterBase &rewriter,
+                        IREE::GPU::ValueBarrierOp barrierA,
+                        IREE::GPU::ValueBarrierOp barrierB) {
+  // Both barriers need to have either tensor semantics or vector semantics.
+  if (barrierA.hasTensorSemantics() && !barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+  if (!barrierA.hasTensorSemantics() && barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+
+  // We assume barrierA is always before barrierB.
+  if (barrierB->isBeforeInBlock(barrierA)) {
+    std::swap(barrierA, barrierB);
+  }
+
+  // barrierA and barrierB are in the same block.
+  Block *block = barrierA->getBlock();
+
+  auto sliceFilter = [&block](Operation *candidate) -> bool {
+    if (candidate->getBlock() != block) {
+      return false;
+    }
+    if (candidate == block->getTerminator()) {
+      // Do not move the terminator.
+      return false;
+    }
+    return true;
+  };
+
+  // Find the combined backward slice of barrierA and barrierB and try
+  // to move it before barrierA (before both the barriers).
+  BackwardSliceOptions bOptions;
+  bOptions.filter = sliceFilter;
+  SetVector<Operation *> backwardSliceA;
+  SetVector<Operation *> backwardSliceB;
+  getBackwardSlice(barrierA, &backwardSliceA, bOptions);
+  getBackwardSlice(barrierB, &backwardSliceB, bOptions);
+  backwardSliceA.insert(backwardSliceB.begin(), backwardSliceB.end());
+  // Move the backward slice before barrierA.
+  if (failed(moveBackwardSliceBeforeOp(rewriter, backwardSliceA, barrierA))) {
+    return failure();
+  }
+
+  // Find the combined forward slice of barrierA and barrierB and try to
+  // move it after barrierB (after both the barriers).
+  ForwardSliceOptions fOptions;
+  fOptions.filter = sliceFilter;
+  SetVector<Operation *> forwardSliceA;
+  SetVector<Operation *> forwardSliceB;
+  getForwardSlice(barrierA, &forwardSliceA, fOptions);
+  getForwardSlice(barrierB, &forwardSliceB, fOptions);
+  forwardSliceA.insert(forwardSliceB.begin(), forwardSliceB.end());
+  // Move the forward slice after barrierB.
+  if (failed(moveForwardSliceAfterOp(rewriter, forwardSliceA, barrierB))) {
+    return failure();
+  }
+
+  // We add the new barrier after both the barriers (it is always better
+  // to sink barriers).
+  rewriter.setInsertionPointAfter(barrierB);
```

Comment:
General good practice to have an `OpBuilder::InsertionGuard` for every method that is modifying the insertion point.